packages/run-protocol/src/vpool-xyk-amm/stableSwapCurve.js

```diff
@@ -0,0 +1,388 @@
+  assert(
+    inputReserve.value > 0n,
+    X`inputReserve ${inputReserve.value} must be positive`,
+  );
+  assert(
+    outputReserve.value > 0n,
+    X`outputReserve ${outputReserve.value} must be positive`,
+  );
+  // Fee ratio calculation
+  const feeCutRatio = makeRatio(
+    BASIS_POINTS - feeBasisPoints,
+    inputAmount.brand,
+    BASIS_POINTS,
+    inputAmount.brand,
+  );
+  // Fee ratio multiplied by inputAmount to get inputAmount After fee cut
+  let inputAmountAfterFeeCut = floorMultiplyBy(inputAmount, feeCutRatio);
+  // Normalizing input amount according to pool value
+  inputAmountAfterFeeCut = AmountMath.make(
+    inputAmountAfterFeeCut.brand,
+    inputAmountAfterFeeCut.value * BASIS_POINTS,
+  );
+  const basisRatio = makeRatio(BASIS_POINTS * 100n, inputAmount.brand);
+  let inputAmountWithoutFeeCut = floorMultiplyBy(inputAmount, basisRatio);
+  // Normalizing the poolValue according to Basis_Points
+  let poolAmountsInBasisPoints = poolAmounts.map(amount => {
+    return floorMultiplyBy(
+      amount,
+      makeRatio(BASIS_POINTS * 100n, amount.brand),
+    );
+  });
+  const poolValues = poolAmountsInBasisPoints.map(amount => amount.value);
+  return { inputAmountAfterFeeCut, inputAmountWithoutFeeCut, poolValues };
+};
+/**
+ * Recompute the pool values after a swap is performed
+ * @param {bigint[]} poolValues - Array of amounts of each asset.Which is
+ *                                passed from the contract.
+ * @param {number} tokenIndexFrom - index of amount of inputReserve
+ *                                  in the reserves array.
+ * @param {number} tokenIndexTo - index of amount of outputReserve
+ *                                in the reserves array.
+ * @param {bigint} inputTokenValue - the value of swap in token passed in.
+ * @param {bigint} outputTokenValue - the value of swap out token returned
+ *                                    in exchange
+ * @returns {bigint[]} output - recomputed pool values
+ */
+const updatePoolValues = (
+  poolValues,
+  tokenIndexFrom,
+  inputTokenValue,
+  tokenIndexTo,
+  outputTokenValue,
+) => {
+  poolValues[tokenIndexFrom] -= inputTokenValue;
+  poolValues[tokenIndexTo] += outputTokenValue;
+  return poolValues;
+};
+
+/**
+ * Computes the Stable Swap invariant (D).
+ * @param {bigint[]} poolValues - Array of amounts of each asset.Which is
+ * passed from the contract.
+ * @returns {bigint} d - the current price, in value form
+ *
+ */
+export const getD = poolValues => {
+  const N_COINS = poolValues.length;
+  // sum_x - Sum of all poolValues.
+  const sum_x = poolValues.reduce((prev, cur) => prev + cur, 0n);
+  if (sum_x === 0n) {
+    return 0n;
+  }
+  let d_prev = 0n;
+  let d = sum_x;
+  const Ann = A * N_COINS * N_COINS;
+  for (let i = 0; i < maxLoopLimit; i++) {
+    let dp = d;
+    // prod - product of all poolvalues
+    // dp = D^(n+1)/n^n(prod)
+    for (let j = 0; j < N_COINS; j++) {
+      dp = (dp * d) / (poolValues[j] * Nat(N_COINS));
+    }
```

Comment:
```suggestion
  // prod:: product of all pool values
  const prod = poolValues.reduce((p, c) => p * c, 1n);
  for (let i = 0; i < maxLoopLimit; i++) {
    // dp = D^(n+1)/n^n(prod)
    const dp = makeRatio(
      d ** (nCoins ** 1n),
      brand,
      (nCoins ** nCoins) * prod,
   );
```

...

Maybe rather than a Ratio, we can just keep separate variables for numerator and denominator, and use those in the calculation of d below.

Reply:
alright